bug: Non-Deterministic Test `KafkaProxyReconcilerIT` on MicroShift

A test failed non-deterministically in GitHub Actions:

## failed test

Job: Operator integration & system tests on MicroShift

Test: `io.kroxylicious.kubernetes.operator.reconciler.kafkaproxy.KafkaProxyReconcilerIT.deploymentChecksumUpdatesWhenOpenshiftRouteHostAssigned`

Build: https://github.com/kroxylicious/kroxylicious/actions/runs/24218917776/job/70707372281

## stack trace

```
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://api.crc.testing:6443/apis/route.openshift.io/v1/namespaces/kafkaproxyreconcilerit-deploymentchecksumupdateswhenopens-21db4/routes/bar-bootstrap/status. Message: Operation cannot be fulfilled on routes.route.openshift.io "bar-bootstrap": the object has been modified; please apply your changes to the latest version and try again. Received status: Status(apiVersion=v1, code=409, details=StatusDetails(causes=[], group=route.openshift.io, kind=routes, name=bar-bootstrap, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Operation cannot be fulfilled on routes.route.openshift.io "bar-bootstrap": the object has been modified; please apply your changes to the latest version and try again, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Conflict, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:205)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:507)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:524)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:419)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:397)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handlePatch(BaseOperation.java:776)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.lambda$patch$2(HasMetadataOperation.java:231)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:236)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patchStatus(HasMetadataOperation.java:241)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patchStatus(HasMetadataOperation.java:44)
	at io.kroxylicious.kubernetes.operator.LocallyRunningOperatorRbacHandler$1.patchStatus(LocallyRunningOperatorRbacHandler.java:231)
	at io.kroxylicious.kubernetes.operator.reconciler.kafkaproxy.KafkaProxyReconcilerIT.deploymentChecksumUpdatesWhenOpenshiftRouteHostAssigned(KafkaProxyReconcilerIT.java:990)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://api.crc.testing:6443/apis/route.openshift.io/v1/namespaces/kafkaproxyreconcilerit-deploymentchecksumupdateswhenopens-21db4/routes/bar-bootstrap/status. Message: Operation cannot be fulfilled on routes.route.openshift.io "bar-bootstrap": the object has been modified; please apply your changes to the latest version and try again. Received status: Status(apiVersion=v1, code=409, details=StatusDetails(causes=[], group=route.openshift.io, kind=routes, name=bar-bootstrap, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Operation cannot be fulfilled on routes.route.openshift.io "bar-bootstrap": the object has been modified; please apply your changes to the latest version and try again, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Conflict, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:642)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:622)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:582)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:549)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:141)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.utils.AsyncUtils.lambda$retryWithExponentialBackoff$3(AsyncUtils.java:91)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:844)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)

```

--- a/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconcilerIT.java
+++ b/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconcilerIT.java
@@ -984,7 +984,7 @@ public class KafkaProxyReconcilerIT {
 
         // The test is racing with the Ingress controller. We don't know if the ingress
         // controller will have updated the resource or not yet. We make an update to
-        // the Route that are certain will be different to that made by the controller
+        // the Route that we are certain will be different to that made by the controller
         // in order to be certain an update is made.
         testActor.resources(Route.class).resource(bootstrapRoute).editStatus(r -> new RouteBuilder(bootstrapRoute)
                 .withNewStatus()

--- a/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconcilerIT.java
+++ b/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaproxy/KafkaProxyReconcilerIT.java
@@ -976,24 +976,28 @@ public class KafkaProxyReconcilerIT {
         updateStatusObservedGeneration(testActor.create(cluster));
 
         String bootstrapRouteName = name(cluster) + "-bootstrap";
-        AWAIT.alias("bootstrap route created").untilAsserted(() -> assertThat(testActor.get(Route.class, bootstrapRouteName)).isNotNull());
+        Route bootstrapRoute = AWAIT.alias("bootstrap route created").until(() -> testActor.get(Route.class, bootstrapRouteName), Objects::nonNull);
 
-        String initialChecksum = deploymentPodTemplateChecksum(proxy);
+        var observedChecksum = deploymentPodTemplateChecksum(proxy);
 
-        // When - OpenShift assigns a host to the bootstrap route
-        Route bootstrapRoute = testActor.get(Route.class, bootstrapRouteName);
-        Route routeWithHost = new RouteBuilder(bootstrapRoute)
+        // When
+
+        // The test is racing with the Ingress controller. We don't know if the ingress
+        // controller will have updated the resource or not yet. We make an update to
+        // the Route that are certain will be different to that made by the controller
+        // in order to be certain an update is made.
+        testActor.resources(Route.class).resource(bootstrapRoute).editStatus(r -> new RouteBuilder(bootstrapRoute)
                 .withNewStatus()
-                .addNewIngress().withHost(bootstrapRouteName + ".apps-crc.testing").endIngress()
+                .addNewIngress().withHost("different.invalid").endIngress()
                 .endStatus()
-                .build();
-        testActor.patchStatus(routeWithHost);
+                .build());
 
         // Then - deployment pod template checksum should change so the proxy is restarted
         AWAIT.alias("deployment checksum updated after route host assignment")
-                .untilAsserted(() -> assertThat(deploymentPodTemplateChecksum(proxy)).isNotEqualTo(initialChecksum));
+                .untilAsserted(() -> assertThat(deploymentPodTemplateChecksum(proxy)).isNotEqualTo(observedChecksum));
     }
 
+    @Nullable
     private String deploymentPodTemplateChecksum(KafkaProxy proxy) {
         var deployment = testActor.get(Deployment.class, ProxyDeploymentDependentResource.deploymentName(proxy));
         assertThat(deployment).isNotNull();

